File: python/agentlens/integrations/openai.py
"""
Drop-in wrapper for the OpenAI Python client.

Usage::

    import openai
    from agentlens import Profiler
    from agentlens.integrations.openai import ProfiledOpenAI

    profiler = Profiler("my-agent")
    client = ProfiledOpenAI(openai.OpenAI(api_key="..."), profiler=profiler)

    # Use exactly like the standard openai client
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": "Hello!"}],
    )
"""

from __future__ import annotations

from typing import Any, Iterator, Optional

from ..profiler import CallType, ProfiledCall, Profiler, TokenUsage


class _ProfiledChatCompletions:
    def __init__(self, inner: Any, profiler: Profiler) -> None:
        self._inner = inner
        self._profiler = profiler

    def create(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        stream = kwargs.get("stream", False)
        name = f"openai.chat/{model}"

        if stream:
            return self._create_stream(name, *args, **kwargs)

        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = self._inner.create(*args, **kwargs)
            usage = TokenUsage.from_openai_usage(getattr(response, "usage", None))
            inferred_model = getattr(response, "model", model)
            call.finish(success=True, token_usage=usage, model=inferred_model)
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)

    def _create_stream(self, name: str, *args: Any, **kwargs: Any) -> Any:
        """Wrap a streaming response — records the call after the stream is consumed."""
        model = kwargs.get("model", "unknown")
        call = self._profiler._start_call(name, CallType.LLM, model=model)

        try:
            stream = self._inner.create(*args, **kwargs)
        except Exception as exc:
            call.finish(success=False, error=exc)
            self._profiler._record(call)
            raise

        return _StreamWrapper(stream, call, self._profiler)

    async def acreate(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        name = f"openai.chat/{model}"
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = await self._inner.acreate(*args, **kwargs)
            usage = TokenUsage.from_openai_usage(getattr(response, "usage", None))
            call.finish(success=True, token_usage=usage, model=getattr(response, "model", model))
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)


class _StreamWrapper:
    """Thin iterator wrapper that closes the profiled call when the stream ends."""

    def __init__(self, stream: Any, call: ProfiledCall, profiler: Profiler) -> None:
        self._stream = stream
        self._call = call
        self._profiler = profiler

    def __iter__(self) -> Iterator[Any]:
        try:
            for chunk in self._stream:
                yield chunk
            self._call.finish(success=True)
        except Exception as exc:
            self._call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(self._call)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._stream, item)


class _ProfiledCompletions:
    """Wraps legacy openai.completions (text completion endpoint)."""

    def __init__(self, inner: Any, profiler: Profiler) -> None:
        self._inner = inner
        self._profiler = profiler

    def create(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        name = f"openai.completion/{model}"
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = self._inner.create(*args, **kwargs)
            usage = TokenUsage.from_openai_usage(getattr(response, "usage", None))
            call.finish(success=True, token_usage=usage)
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)


class _ProfiledEmbeddings:
    def __init__(self, inner: Any, profiler: Profiler) -> None:
        self._inner = inner
        self._profiler = profiler

    def create(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "text-embedding-3-small")
        name = f"openai.embeddings/{model}"
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = self._inner.create(*args, **kwargs)
            usage = TokenUsage.from_openai_usage(getattr(response, "usage", None))
            call.finish(success=True, token_usage=usage)
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)


class ProfiledOpenAI:
    """
    Wraps an ``openai.OpenAI`` (or ``openai.AzureOpenAI``) instance and
    automatically profiles every API call.

    All attributes not explicitly wrapped are forwarded transparently to the
    underlying client, so you can drop this in without any other code changes.
    """

    def __init__(self, client: Any, profiler: Optional[Profiler] = None) -> None:
        self._client = client
        self._profiler = profiler or Profiler("openai")
        self.chat = _ProfiledChat(client.chat, self._profiler)
        self.completions = _ProfiledCompletions(client.completions, self._profiler)  # type: ignore[attr-defined]
        self.embeddings = _ProfiledEmbeddings(client.embeddings, self._profiler)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._client, item)

    @property
    def profiler(self) -> Profiler:
        return self._profiler


class _ProfiledChat:
    def __init__(self, inner: Any, profiler: Profiler) -> None:
        self._inner = inner
        self.completions = _ProfiledChatCompletions(inner.completions, profiler)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._inner, item)

File: typescript/src/integrations/openai.ts
/**
 * AgentLens — Drop-in wrapper for the OpenAI TypeScript SDK.
 *
 * @example
 * ```ts
 * import OpenAI from "openai";
 * import { Profiler } from "../profiler";
 * import { ProfiledOpenAI } from "./openai";
 *
 * const profiler = new Profiler("my-agent");
 * const openai = new ProfiledOpenAI(new OpenAI({ apiKey: "..." }), profiler);
 *
 * // Works exactly like the normal client
 * const res = await openai.chat.completions.create({
 *   model: "gpt-4o",
 *   messages: [{ role: "user", content: "Hello!" }],
 * });
 * ```
 */

import { Profiler, TokenUsage } from "../profiler";

type AnyFn = (...args: unknown[]) => unknown;

function extractUsage(response: unknown): Partial<TokenUsage> | null {
  if (!response || typeof response !== "object") return null;
  const r = response as Record<string, unknown>;
  if (r["usage"] && typeof r["usage"] === "object") {
    const u = r["usage"] as Record<string, unknown>;
    if ("prompt_tokens" in u) {
      return {
        promptTokens: Number(u["prompt_tokens"]) || 0,
        completionTokens: Number(u["completion_tokens"]) || 0,
        totalTokens: Number(u["total_tokens"]) || 0,
      };
    }
  }
  return null;
}

class ProfiledChatCompletions {
  constructor(
    private readonly inner: { create: AnyFn },
    private readonly profiler: Profiler,
  ) {}

  async create(params: Record<string, unknown>, opts?: unknown): Promise<unknown> {
    const model = String(params["model"] ?? "unknown");
    const name = `openai.chat/${model}`;
    const call = this.profiler.startCall(name, "llm", model);

    try {
      const response = await (this.inner.create as (p: unknown, o?: unknown) => Promise<unknown>)(
        params,
        opts,
      );
      const usage = extractUsage(response);
      const inferredModel =
        response && typeof response === "object" && "model" in (response as object)
          ? String((response as Record<string, unknown>)["model"])
          : model;
      this.profiler.endCall(call, {
        success: true,
        tokenUsage: usage ?? undefined,
        model: inferredModel,
      });
      return response;
    } catch (err) {
      this.profiler.endCall(call, {
        success: false,
        error: err instanceof Error ? err : new Error(String(err)),
      });
      throw err;
    }
  }
}

class ProfiledChat {
  readonly completions: ProfiledChatCompletions;

  constructor(innerChat: { completions: { create: AnyFn } }, profiler: Profiler) {
    this.completions = new ProfiledChatCompletions(innerChat.completions, profiler);
  }
}

class ProfiledEmbeddings {
  constructor(
    private readonly inner: { create: AnyFn },
    private readonly profiler: Profiler,
  ) {}

  async create(params: Record<string, unknown>, opts?: unknown): Promise<unknown> {
    const model = String(params["model"] ?? "text-embedding-3-small");
    const name = `openai.embeddings/${model}`;
    const call = this.profiler.startCall(name, "llm", model);

    try {
      const response = await (this.inner.create as (p: unknown, o?: unknown) => Promise<unknown>)(
        params,
        opts,
      );
      const usage = extractUsage(response);
      this.profiler.endCall(call, { success: true, tokenUsage: usage ?? undefined });
      return response;
    } catch (err) {
      this.profiler.endCall(call, {
        success: false,
        error: err instanceof Error ? err : new Error(String(err)),
      });
      throw err;
    }
  }
}

/**
 * Drop-in wrapper for an `OpenAI` client instance.
 * All other attributes are forwarded transparently.
 */
export class ProfiledOpenAI {
  readonly chat: ProfiledChat;
  readonly embeddings: ProfiledEmbeddings;

  constructor(
    private readonly client: {
      chat: { completions: { create: AnyFn } };
      embeddings: { create: AnyFn };
      [key: string]: unknown;
    },
    public readonly profiler: Profiler = new Profiler("openai"),
  ) {
    this.chat = new ProfiledChat(client.chat, profiler);
    this.embeddings = new ProfiledEmbeddings(client.embeddings, profiler);
  }

  // Forward any other attributes to the underlying client
  [key: string]: unknown;
}
